### ISSUE TITLE
Allow overriding optimiser step / zero grad action 

### ISSUE THREAD

**Issue by soraxas** [williamFalcon/pytorch-lightning#106]
Currently the logic of `optimiser.step()` and `optimiser.zero_grad()` are hard coded in the trainer, but sometimes it would be benificial to NOT `zero_grad()` or perform it at an arbitrary iteration (e.g. For RNN).
This is also related to #29 of implementing GAN in lightning.

**Comment by williamFalcon** [comment]
@soraxas good suggestion. Can you propose an  alternative set up? would love to see other options for generalizing.   

Maybe some hooks or trainer flags?

**Comment by williamFalcon** [comment]
would #107 work?

**Comment by williamFalcon** [comment]
@soraxas. this fix is on master now. try it out and let me know if it solves your issue. 

use case 2 for a gan

**Comment by AlphabetMan** [comment]
GANs are getting more and more complex. There might be more than 2 backward calls in one training step. I suggest to define optional methods training_substep_1 or _2, _3 etc. which consist of these forward-backward passes and returns a loss. Substeps would be contained within training_step there all losses would be accumulated for logging.

**Comment by williamFalcon** [comment]
@AlphabetMan can you give me a pseudocode example? this pattern isn't limited to 2 optimizers, just used 2 here for illustration.

**Comment by williamFalcon** [comment]
docs here: https://williamfalcon.github.io/pytorch-lightning/Trainer/hooks/#optimizer_step

**Comment by AlphabetMan** [comment]
I have been thinking. Optimizer indexing seems to be inevitable. However, I don't like the idea of skipping batches, even though it might not affect quality of the results. I have came up with this.  This gets complicated rather quickly.

`class CoolModel(pl.LightningModule):

    def __init__(self):
        super(CoolModel, self).__init__()
        # not the best model...
        self.discriminator = torch.nn.Linear(28 * 28, 10)
        self.generator = torch.nn.Linear(28 * 28, 10)
        self.loss1 = nn.MSELoss()
        self.loss2 = nn.BCELoss()
    
    # becomes irrelevant
#     def forward(self, x):
#         return torch.relu(self.l1(x.view(x.size(0), -1)))
    @MaybeSomeDecorator
    def training_substep_1(self, batch):
        # on real
        output1 = self.discriminator(batch)
        loss1 = self.loss2(output1, torch.ones_like(output1))
        # on fake
        fake_image = self.generator(torch.randn(8,28*16))
        output2 = self.discriminator(fake_image.detach())
        loss2 = self.loss2(output2, torch.zeros_like(output2))
        loss = loss1 + loss2
        return {'loss': loss, 'optimizer_idx': 0, 'tensor': fake_image}
    
    def training_step(self, batch, batch_nb):
        intermediate = self.training_substep_1(batch) # backward is called for discriminator
        output1 = self.discriminator
        loss = self.loss2(output1, torch.ones_like(output1)) 
        return {'loss': loss, 'optimizer_idx': 1} # backward is called for generator

    def configure_optimizers(self):
        return [torch.optim.Adam(self.discriminator.parameters(), lr=0.02), torch.optim.Adam(self.generator.parameters(), lr=0.02)]`

**Comment by williamFalcon** [comment]
The current approach live on 0.4.5 is:

#### Training_step    
```python
def training_step(self, batch, batch_nb, optimizer_i):
    if optimizer_i == 0:
        # do generator stuff
    if optimizer_i == 1:
        # do discriminator stuff
```

#### Control optimizer behavior     
```python
def optimizer_step(self, epoch_nb, batch_nb, optimizer, optimizer_i):
        """
        Do something instead of the standard optimizer behavior
        :param epoch_nb:
        :param batch_nb:
        :param optimizer:
        :param optimizer_i:
        :return:
        """
        # update generator opt every 2 steps
        if optimizer_i == 0:
            if batch_nb % 2 == 0 :
                optimizer.step()
                optimizer.zero_grad()
       
        # update discriminator opt every 4 steps
        if optimizer_i == 1:
            if batch_nb % 4 == 0 :
                optimizer.step()
                optimizer.zero_grad()

        # update other opt every 6 steps
        if optimizer_i == 2:
            if batch_nb % 6 == 0 :
                optimizer.step()
                optimizer.zero_grad()
```

And of course, return 2 optimizers
```python    
def configure_optimizers(self):
    # REQUIRED
    return [torch.optim.Adam(self.discriminator.parameters(), lr=0.02), torch.optim.Adam(self.generator.parameters(), lr=0.02)]`
```

**Comment by williamFalcon** [comment]
@AlphabetMan thoughts on this approach currently available?

**Comment by AlphabetMan** [comment]
So if there are multiple optimizers trainer calls sequentially:
`training_step(self, batch, batch_nb, optimizer_i=0)
optimizer_step(self, epoch_nb, batch_nb, optimizer, optimizer_i=0)
training_step(self, batch, batch_nb, optimizer_i=1)
optimizer_step(self, epoch_nb, batch_nb, optimizer, optimizer_i=1)
training_step(self, batch, batch_nb, optimizer_i=2)
optimizer_step(self, epoch_nb, batch_nb, optimizer, optimizer_i=2)`
etc, until run out of optimizers.
Do I understand this correctly?
If yes, then it works. It's just that discriminator is updated with detached tensor from generated fake image. Usually, then the same tensor is passed through discriminator again to update generator. But in this implementation, to update generator image has to be generated again. It's not effective. Unless of course, we can save that generated tensor in unused attribute in every step, for example:
`self.temporary_tensor,`

**Comment by williamFalcon** [comment]
yup... that's what happens right now!

**Comment by williamFalcon** [comment]
@AlphabetMan would be super helpful if you put together a simple GAN example to:
A. make sure it makes sense.
B. have as an official example on doing GAN training with Lightning

**Comment by williamFalcon** [comment]
@AlphabetMan i added a template gan example to show how it works.

**Comment by williamFalcon** [comment]
![image](https://user-images.githubusercontent.com/3640001/63021683-213ffe00-be6f-11e9-83cc-3bb0b518edb8.png)

**Comment by AlphabetMan** [comment]
Very nice. There are multiple ways to define training step, any network could be updated before the other, but this example is clear and simple. GAN's really benefit from distributed training, so this framework is really helpful. I think any tutorial on deep nets on pytorch website should be implementable here.
